File: packages/flutter-im-ui/lib/src/models/message_grouping.dart
import 'message_data.dart';

enum FlareMessageGroupPosition { single, first, middle, last }

enum FlareAvatarPlacement { leading, trailing }

class FlareMessageRowPresentation {
  const FlareMessageRowPresentation({
    this.showAvatar = false,
    this.reserveAvatarSpace = false,
    this.showSenderName = false,
    this.avatarPlacement = FlareAvatarPlacement.leading,
  });

  final bool showAvatar;
  final bool reserveAvatarSpace;
  final bool showSenderName;
  final FlareAvatarPlacement avatarPlacement;
}

bool flareMessagesShareGroup(
  FlareMessageData? previous,
  FlareMessageData? current, {
  int timeGapMs = 300000,
}) {
  // Notices (system lines, recalled messages) stand alone and end a sender's run.
  if (previous == null ||
      current == null ||
      previous.isSystem ||
      current.isSystem ||
      previous.isRecalled ||
      current.isRecalled)
    return false;
  if (previous.senderId != current.senderId) return false;
  if (previous.sentAtMs <= 0 || current.sentAtMs <= 0) return true;
  return (current.sentAtMs - previous.sentAtMs).abs() <= timeGapMs;
}

FlareMessageGroupPosition flareMessageGroupPosition(
  List<FlareMessageData> messages,
  int index, {
  int timeGapMs = 300000,
}) {
  final current = messages[index];
  if (current.isSystem || current.isRecalled) {
    return FlareMessageGroupPosition.single;
  }
  final previous = index > 0 ? messages[index - 1] : null;
  final next = index + 1 < messages.length ? messages[index + 1] : null;
  final joinsPrevious = flareMessagesShareGroup(
    previous,
    current,
    timeGapMs: timeGapMs,
  );
  final joinsNext = flareMessagesShareGroup(
    current,
    next,
    timeGapMs: timeGapMs,
  );
  if (!joinsPrevious && !joinsNext) return FlareMessageGroupPosition.single;
  if (!joinsPrevious) return FlareMessageGroupPosition.first;
  if (!joinsNext) return FlareMessageGroupPosition.last;
  return FlareMessageGroupPosition.middle;
}

FlareMessageRowPresentation flareMessageRowPresentation({
  required FlareMessageData message,
  required FlareMessageGroupPosition position,
  required String currentUserId,
  required bool groupConversation,
  bool showIncomingAvatar = true,
  bool showSelfAvatar = false,
  bool showGroupSenderName = true,
}) {
  final self = message.senderId == currentUserId;
  final edge =
      position == FlareMessageGroupPosition.single ||
      position == FlareMessageGroupPosition.first;
  return FlareMessageRowPresentation(
    showAvatar: (self ? showSelfAvatar : showIncomingAvatar) && edge,
    reserveAvatarSpace: self ? showSelfAvatar : showIncomingAvatar,
    showSenderName: !self && groupConversation && showGroupSenderName && edge,
    avatarPlacement: self
        ? FlareAvatarPlacement.trailing
        : FlareAvatarPlacement.leading,
  );
}

File: packages/vue-im-ui/src/shared/contracts/message-grouping.ts
import type { MessageLike } from "./messageRow";
import type { FlareConversationKind } from "./conversation";

export type FlareMessageGroupPosition = "single" | "first" | "middle" | "last";

export interface FlareMessageRowPresentation {
  showAvatar: boolean;
  reserveAvatarSpace: boolean;
  showSenderName: boolean;
  avatarPlacement: "leading" | "trailing";
}

export interface FlareMessageGroupingOptions {
  timeGapMs?: number;
  showIncomingAvatar?: boolean;
  showSelfAvatar?: boolean;
  showGroupSenderName?: boolean;
}

const DEFAULT_TIME_GAP_MS = 5 * 60_000;

export function isSystemMessageForGrouping(message: MessageLike | undefined): boolean {
  const type = message?.content?.contentType;
  return type === "system" || type === "notification";
}

export function messagesShareGroup(
  previous: MessageLike | undefined,
  current: MessageLike | undefined,
  timeGapMs = DEFAULT_TIME_GAP_MS,
): boolean {
  if (!previous || !current) return false;
  if (isSystemMessageForGrouping(previous) || isSystemMessageForGrouping(current)) return false;
  const senderId = previous.senderId?.trim();
  if (!senderId || senderId !== current.senderId?.trim()) return false;
  const previousTs = Number(previous.createdAt || previous.clientCreatedAt || 0);
  const currentTs = Number(current.createdAt || current.clientCreatedAt || 0);
  return !previousTs || !currentTs || Math.abs(currentTs - previousTs) <= timeGapMs;
}

export function messageGroupPosition(
  messages: readonly MessageLike[],
  index: number,
  timeGapMs = DEFAULT_TIME_GAP_MS,
): FlareMessageGroupPosition {
  const current = messages[index];
  if (!current || isSystemMessageForGrouping(current)) return "single";
  const joinedPrevious = messagesShareGroup(messages[index - 1], current, timeGapMs);
  const joinedNext = messagesShareGroup(current, messages[index + 1], timeGapMs);
  if (!joinedPrevious && !joinedNext) return "single";
  if (!joinedPrevious) return "first";
  if (!joinedNext) return "last";
  return "middle";
}

export function messageRowPresentation(
  message: MessageLike,
  position: FlareMessageGroupPosition,
  currentUserId: string,
  conversationKind: FlareConversationKind | undefined,
  options: FlareMessageGroupingOptions = {},
): FlareMessageRowPresentation {
  const self = message.senderId === currentUserId;
  const edge = position === "single" || position === "first";
  const showAvatar = self
    ? Boolean(options.showSelfAvatar) && edge
    : (options.showIncomingAvatar ?? true) && edge;
  return {
    showAvatar,
    reserveAvatarSpace: self ? Boolean(options.showSelfAvatar) : (options.showIncomingAvatar ?? true),
    showSenderName:
      !self
      && conversationKind === "group"
      && (options.showGroupSenderName ?? true)
      && edge,
    avatarPlacement: self ? "trailing" : "leading",
  };
}
